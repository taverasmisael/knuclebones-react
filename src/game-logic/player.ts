export type PlayerId = string
export type PlayerBoardPosition = 'top' | 'bottom'
export const PlayerBoardPosition = {
  TOP: 'top' as const,
  BOTTOM: 'bottom' as const,
}
export type PairOfPlayers = [PlayerId, PlayerId]
