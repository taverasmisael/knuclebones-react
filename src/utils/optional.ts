interface toJSON {
	toJSON(): string;
	toString(): string;
}
interface Maybe<T> extends toJSON {
	readonly _type: "SOME" | "NONE";
	value?: T;
	unwrap(): T;
	map<R>(fn: (value: T) => R): Optional<R>;
}

export interface Some<T> extends Maybe<T> {
	readonly _type: "SOME";
	value: T;
}
export interface None extends Maybe<never> {
	readonly _type: "NONE";
}
export type Optional<T> = Some<T> | None;
type NonNullable = string | number | boolean | symbol | object;

export function Optional<T>(value?: T): Optional<T> {
	if (value == null) {
		return None();
	}
	return Some(value);
}

export function isNone<T>(value: Optional<T>): value is None {
	return value._type === "NONE";
}

export function isSome<T>(value: Optional<T>): value is Some<T> {
	return value._type === "SOME";
}

export function Some<T extends NonNullable>(value: T): Some<T> {
	return {
		_type: "SOME",
		value,
		unwrap: () => value,
		map: (fn) => Optional(fn(value)),
		toJSON: () => `Some(${JSON.stringify(value)})`,
		toString: () => `Some(${JSON.stringify(value)})`,
	};
}

export function None(): None {
	return {
		_type: "NONE",
		unwrap: () => {
			throw new Error("Cannot unwrap None");
		},
		map: () => None(),
		toJSON: () => "None()",
		toString: () => "None()",
	};
}

// helpers
export function unwrapOr<T>(defaultValue: T, value: Optional<T>): T {
	if (isNone(value)) {
		return defaultValue;
	}
	return value.unwrap();
}

export function contains<T>(value: Optional<T>, search: T): boolean {
	if (isNone(value)) {
		return false;
	}
	return value.value === search;
}

export function flat<T>(value: Optional<Optional<T>>): Optional<T> {
	if (isNone(value)) {
		return None();
	}
	return value.value;
}

export function flatMap<T, R>(value: Optional<T>, fn: (value: T) => Optional<R>): Optional<R> {
	if (isNone(value)) {
		return None();
	}
	return fn(value.value);
}
