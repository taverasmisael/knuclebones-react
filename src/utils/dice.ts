import { DiceValue } from "../game-logic/dice";

// min and max included
export function rnd<T extends number>(min: T, max: T): T {
	if (max < min) throw new Error("max cannot be less than min");
	return Math.floor(Math.random() * (max - min + 1) + min) as T;
}

export function rollDice(): DiceValue {
	return rnd<DiceValue>(1, 6);
}
