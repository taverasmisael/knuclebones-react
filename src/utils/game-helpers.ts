import produce from 'immer'
import { BoardCoordinate, makeMoveOnBoard, Board } from '../game-logic/board'
import { DiceValue } from '../game-logic/dice'
import {
  GameState,
  isGameRunning,
  createNewGameOver,
} from '../game-logic/game-status'
import { PlayerBoardPosition } from '../game-logic/player'
import { isNone, Some, isSome } from './optional'

export const makeMove = (
  player: PlayerBoardPosition,
  coordinate: BoardCoordinate,
  diceValue: DiceValue,
  gameStatus: GameState
) =>
  produce(gameStatus, (gs: GameState) => {
    if (!isGameRunning(gs)) return
    gs.board = gs.board.map((b) =>
      makeMoveOnBoard(b, player, coordinate, diceValue)
    ) as Some<Board>
    gs.currentPlayer = Some(
      player === PlayerBoardPosition.TOP
        ? PlayerBoardPosition.BOTTOM
        : PlayerBoardPosition.TOP
    )
  })

export const canPlayerMove = (
  player: PlayerBoardPosition,
  gs: GameState
): boolean => {
  if (!isGameRunning(gs)) return false
  if (isNone(gs.currentPlayer)) return false
  if (gs.currentPlayer.unwrap() !== player) return false
  return true
}

export const checkForWinner = (gs: GameState): GameState => {
  if (!isGameRunning(gs)) return gs
  if (isNone(gs.board)) return gs
  if (isNone(gs.players)) return gs
  const board = gs.board.unwrap()
  const topFull = board.top.every((row) => row.every((cell) => isSome(cell)))
  const bottomFull = board.bottom.every((row) =>
    row.every((cell) => isSome(cell))
  )
  if (topFull || bottomFull) {
    return createNewGameOver(
      gs,
      topFull ? PlayerBoardPosition.TOP : PlayerBoardPosition.BOTTOM
    )
  }
  return gs
}
