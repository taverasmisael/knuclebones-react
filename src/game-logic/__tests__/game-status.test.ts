import { describe, test, expect } from "vitest";
import { isNone } from "../../utils/optional";
import {
	createNewGameNotStarted,
	createNewGameOver,
	createNewGameRunning,
	isGameRunning,
	isGameOver,
	isGameNotStarted,
} from "../game-status";
import { PlayerBoardPosition } from "../player";

describe("game-status", () => {
	const players = ["player1", "player2"] as [string, string];
	test("creatNewGameNotStarted", () => {
		const actual = createNewGameNotStarted();
		expect(isNone(actual.winner)).toBe(true);
		expect(actual).toMatchSnapshot();
	});

	test("createNewGameRunning", () => {
		const actual = createNewGameRunning(players);
		expect(isNone(actual.winner)).toBe(true);
		expect(actual).toMatchSnapshot();
	});

	test("createNewGameOver (P1 winner)", () => {
		const actual = createNewGameOver(createNewGameRunning(players), PlayerBoardPosition.BOTTOM);
		expect(actual.winner.unwrap()).toBe(PlayerBoardPosition.BOTTOM);
		expect(actual).toMatchSnapshot();
	});

	test("createNewGameOver (P2 winner)", () => {
		const actual = createNewGameOver(createNewGameRunning(players), PlayerBoardPosition.TOP);
		expect(actual.winner.unwrap()).toBe(PlayerBoardPosition.TOP);
		expect(actual).toMatchSnapshot();
	});

	describe("helpers", () => {
		test("isGameNotStarted", () => {
			const actual = isGameNotStarted(createNewGameNotStarted());
			expect(actual).toBe(true);
			expect(isGameNotStarted(createNewGameRunning(players))).toBe(false);
		});

		test("isGameRunning", () => {
			const actual = isGameRunning(createNewGameRunning(players));
			expect(actual).toBe(true);
			expect(isGameRunning(createNewGameNotStarted())).toBe(false);
		});

		test("isGameOver", () => {
			const actual = isGameOver(
				createNewGameOver(createNewGameRunning(players), PlayerBoardPosition.TOP),
			);
			expect(actual).toBe(true);
			expect(isGameOver(createNewGameRunning(players))).toBe(false);
		});
	});
});
