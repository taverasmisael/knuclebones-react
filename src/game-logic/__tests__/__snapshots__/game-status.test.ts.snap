// Vitest Snapshot v1

exports[`game-status > creatNewGameNotStarted 1`] = `
{
  "_type": "GAME_NOT_STARTED",
  "board": "None()",
  "currentPlayer": "None()",
  "players": "None()",
  "winner": "None()",
}
`;

exports[`game-status > createNewGameOver (P1 winner) 1`] = `
{
  "_type": "GAME_OVER",
  "board": "Some({\\"top\\":[[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"1-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-2\\"}]],\\"bottom\\":[[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"1-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-2\\"}]]})",
  "currentPlayer": "None()",
  "players": "Some({\\"top\\":\\"player2\\",\\"bottom\\":\\"player1\\"})",
  "winner": "Some(\\"bottom\\")",
}
`;

exports[`game-status > createNewGameOver (P2 winner) 1`] = `
{
  "_type": "GAME_OVER",
  "board": "Some({\\"top\\":[[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"1-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-2\\"}]],\\"bottom\\":[[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"1-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-2\\"}]]})",
  "currentPlayer": "None()",
  "players": "Some({\\"top\\":\\"player2\\",\\"bottom\\":\\"player1\\"})",
  "winner": "Some(\\"top\\")",
}
`;

exports[`game-status > createNewGameRunning 1`] = `
{
  "_type": "GAME_RUNNING",
  "board": "Some({\\"top\\":[[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"1-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-2\\"}]],\\"bottom\\":[[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"0-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"0-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"1-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"1-2\\"}],[{\\"value\\":\\"None()\\",\\"enabled\\":true,\\"id\\":\\"2-0\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-1\\"},{\\"value\\":\\"None()\\",\\"enabled\\":false,\\"id\\":\\"2-2\\"}]]})",
  "currentPlayer": "Some(\\"bottom\\")",
  "players": "Some({\\"top\\":\\"player2\\",\\"bottom\\":\\"player1\\"})",
  "winner": "None()",
}
`;
